subsys/bluetooth/host/adv.c

```diff
@@ -1703,38 +1747,19 @@ int bt_le_per_adv_set_data(const struct bt_le_ext_adv *adv,
 		return -EINVAL;
 	}
 
-	if (ad_len > BT_HCI_LE_PER_ADV_FRAG_MAX_LEN) {
-		return -EINVAL;
-	}
-
-
-	buf = bt_hci_cmd_create(BT_HCI_OP_LE_SET_PER_ADV_DATA, sizeof(*cp));
-	if (!buf) {
-		return -ENOBUFS;
-	}
-
-	cp = net_buf_add(buf, sizeof(*cp));
-	(void)memset(cp, 0, sizeof(*cp));
-
-	cp->handle = adv->handle;
-
-	/* TODO: If data is longer than what the controller can manage,
-	 * split the data. Read size from controller on boot.
-	 */
-	cp->op = BT_HCI_LE_PER_ADV_OP_COMPLETE_DATA;
-
-	err = set_data_add_complete(cp->data, BT_HCI_LE_PER_ADV_FRAG_MAX_LEN, &d, 1,
-				    &cp->len);
-	if (err) {
-		return err;
+	for (size_t i = 0; i < ad_len; i++) {
+		total_len_bytes += ad[i].data_len + 2;
 	}
 
-	err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_SET_PER_ADV_DATA, buf, NULL);
-	if (err) {
-		return err;
+	if ((total_len_bytes > BT_HCI_LE_PER_ADV_FRAG_MAX_LEN) &&
+	    atomic_test_bit(adv->flags, BT_PER_ADV_ENABLED)) {
+		/* It is not allowed to set periodic advertising data
+		 * in multiple operations while it is running.
+		 */
+		return -EAGAIN;
```

Comment:
If the data is too long, it doesn't really make sense to tell the caller to try again. 

```suggestion
		return -EINVAL;
```